_posts/2020-09-08-continuous-integration-with-unreal-engine-4.md

```diff
@@ -0,0 +1,517 @@
+
+##### Tests
+
+Tests in the CI can run only on Linux - if the application you're developing requires user interaction or if you're using CEF (Chromium Embedded Framework) especially. In fact, in Linux we have Xvfb (X virtual framebuffer) that's a display server implementing the X11 display server protocol. 
+In fact, we can easily customize the Dockerfile of the Linux container to invoke `Xvfb` in the `ENTRYPOINT` and send it to the background. In this way, every container we spawn have its own display running and all the tests written using the Automation Driver can work easily, even if there isn't a physical display present.
+
+To enable this, it's just a matter of adding to the `start.sh` script invoked in the entrypoint the invocation of Xvfb.
+
+```
+#!/usr/bin/env bash
+
+/usr/bin/Xvfb :99 -screen 0 1024x768x24 -ac +extension GLX +render -noreset -nolisten unix &
+export DISPLAY=:99
+/usr/bin/env bash
+````
+
+Once rebuilt the Linux containers, we are ready to define the `test` job.
+
+```yml
+test:
+    stage: tests
+    image: adamrehn/ue4-full:4.25.3
+    script:
+        # The container contains a virtual display (Xvfb)
+        # running, and the DISPLAY variable is already set
+        - ue4 build Development # first build the program (in editor)
+        - ue4 run -ExecCmds="automation RunTests Now MODULE+TO+TEST+PLUS+SEPARATED; quit" -buildmachine -forcelogflush -unattended -nopause -nosplash -log -nullrhi | grep -v "empty engine version" > output.log || true
+        - cat output.log | grep "Test Completed"
+        - if (!(grep -q "Test Completed. Result={Fail" output.log) && (grep -q "TEST COMPLETE. EXIT CODE" output.log)); then exit 0; fi
+        - exit 1
+    only:
+        - merge_requests
+    artifacts:
+        paths:
+            - output.log
+        when: always
+    # Select runner
+    tags:
+        - linux
+        - docker
+        - ue4
+```
+
+This job also produces a `output.log` file that can be inspected in case of crashes of the test suite (it might happen).
+
+**NOTE**: `ue4cli` offers the `ue4 test` command, but when there is no display connected it doesn't work and we have to use the custom command line you can see above in the `ue4 run` invocation.
+
+### Continuous Delivery
+
+We want to give the developers the possibility of downloading a packaged version of the application, in both shipping (optimized) and debug (with debug symbols, `check`, `ensure`, and so on) version and at the same time setup the `gitlab-ci.yml` file in order to generate automatically builds at fixed time intervals (this option can be configured by Gitlab itself using the web interface).
+
+For this reason, instead of directly defining the jobs we define *templates*. These templaces will be then used for both the *manual* and the *scheduled* jobs.
```

Comment:
```suggestion
For this reason, instead of directly defining the jobs we define *templates*. These templates will be then used for both the *manual* and the *scheduled* jobs.
```